File: tools/babel2toucan.py
#!/usr/bin/env python3
"""
Tool: Convert a BabelEdit .babel XML file into a toucan.project JSON format.

Usage:
    python tools/babel2toucan.py d:\locales\qcash-ui.babel > toucan.project

This script mirrors BabelEdit metadata (no actual translation text) into the Toucan JSON schema.
"""
import xml.etree.ElementTree as ET
import json
import sys
from pathlib import Path


def node_text(elem, tag):
    child = elem.find(tag)
    return child.text if child is not None else None


def parse_translations(trans_elem):
    if trans_elem is None:
        return {}
    result = {}
    for tr in trans_elem.findall('translation'):
        lang = node_text(tr, 'language')
        approved_text = node_text(tr, 'approved')
        approved = approved_text.lower() == 'true' if approved_text else False
        result[lang] = {
            'approved': approved
        }
    return result


def parse_concept(concept):
    return {
        'name': node_text(concept, 'name'),
        'description': node_text(concept, 'description') or '',
        'comment': node_text(concept, 'comment') or '',
        'translations': parse_translations(concept.find('translations'))
    }


def parse_folder(node):
    # folder_node => may contain folder_node(s) and concept_node(s)
    name = node_text(node, 'name')
    result = { 'name': name, 'folders': [], 'concepts': [] }
    children = node.find('children')
    if children is None:
        return result
    for c in children:
        if c.tag == 'folder_node':
            result['folders'].append(parse_folder(c))
        elif c.tag == 'concept_node':
            result['concepts'].append(parse_concept(c))
        else:
            # ignore other node types
            continue
    return result


def parse_file(file_node):
    file_name = node_text(file_node, 'name')
    children = file_node.find('children')
    folders = []
    if children is not None:
        for ch in children:
            if ch.tag == 'folder_node':
                folders.append(parse_folder(ch))
            elif ch.tag == 'concept_node':
                # top-level concept in a file
                # Represent as a folderless concept
                folders.append({ 'name': '', 'folders':[], 'concepts':[parse_concept(ch)]})
    return {'name': file_name, 'folders': folders}


def parse_package(pkg):
    name = node_text(pkg, 'name')
    files = []
    children = pkg.find('children')
    if children is not None:
        for child in children:
            if child.tag == 'file_node':
                files.append(parse_file(child))
    return { 'name': name, 'files': files }


def main(argv):
    if len(argv) < 2:
        print('Usage: babel2toucan.py <path/to/project.babel>', file=sys.stderr)
        sys.exit(2)

    p = Path(argv[1])
    if not p.exists():
        print('File not found: ' + str(p), file=sys.stderr)
        sys.exit(2)

    tree = ET.parse(str(p))
    root = tree.getroot()

    project = {
        '$schema': './toucan.project.schema.json',
        'schemaVersion': '1.0.0',
        'projectName': node_text(root, 'filename') or p.stem,
        'beVersion': root.attrib.get('be_version'),
        'framework': node_text(root, 'framework'),
        'languages': [],
        'packages': []
    }

    # Prefer explicit <languages><language><code> entries when present
    langs = []
    langs_node = root.find('languages')
    if langs_node is not None:
        for l in langs_node.findall('language'):
            code = node_text(l, 'code')
            if code:
                langs.append(code)
    else:
        # Fall back to scanning <translation> nodes
        for tr in root.findall('.//translation'):
            lang = node_text(tr, 'language')
            if lang and lang not in langs:
                langs.append(lang)
    project['languages'] = sorted(langs)

    for pkg in root.findall('.//package_node'):
        project['packages'].append(parse_package(pkg))

    # translation_packages mapping (paths to language files)
    tp = []
    tp_node = root.find('translation_packages')
    if tp_node is not None:
        for tpackage in tp_node.findall('translation_package'):
            tname = node_text(tpackage, 'name') or ''
            turls = []
            urls = tpackage.find('translation_urls')
            if urls is not None:
                for url in urls.findall('translation_url'):
                    turls.append({
                        'path': node_text(url, 'path'),
                        'language': node_text(url, 'language')
                    })
            tp.append({'name': tname, 'translationUrls': turls})
    project['translationPackages'] = tp

    # embedded source, primary language, editor and configuration
    est = node_text(root, 'embedded_source_texts')
    project['embeddedSourceTexts'] = (est or 'false').lower() == 'true'
    project['primaryLanguage'] = node_text(root, 'primary_language')

    # editor_configuration => mapping of child text values
    editor_cfg = {}
    ed = root.find('editor_configuration')
    if ed is not None:
        for c in ed:
            # For repeated tags like copy_template, aggregate into a list
            if c.tag == 'copy_template':
                editor_cfg.setdefault('copy_templates', []).append(c.text or '')
            else:
                editor_cfg[c.tag] = c.text
    project['editorConfiguration'] = editor_cfg

    # configuration block, copy children into a dict
    cfg = {}
    conf = root.find('configuration')
    if conf is not None:
        for c in conf:
            cfg[c.tag] = c.text
    project['configuration'] = cfg

    json.dump(project, sys.stdout, indent=2, ensure_ascii=False)


if __name__ == '__main__':
    main(sys.argv)

File: tools/Babel2Toucan.cs
// Usage: dotnet run -- <path/to/project.babel> [> toucan.project]
// Converts a .babel XML project file into the toucan.project JSON format.
// Requires: dotnet-script or compile with `dotnet build` (top-level statements, .NET 10+)

#pragma warning disable CA1869 // Cache and reuse 'JsonSerializerOptions' instances

using System.Text.Json;
using System.Text.Json.Serialization;
using System.Xml.Linq;

if (args.Length < 1)
{
    Console.Error.WriteLine("Usage: Babel2Toucan <path/to/project.babel>");
    return 2;
}

var filePath = args[0];
if (!File.Exists(filePath))
{
    Console.Error.WriteLine($"File not found: {filePath}");
    return 2;
}

var doc = XDocument.Load(filePath);
var root = doc.Root!;

var project = new Dictionary<string, object?>
{
    ["$schema"] = "./toucan.project.schema.json",
    ["schemaVersion"] = "1.0.0",
    ["projectName"] = NodeText(root, "filename") ?? Path.GetFileNameWithoutExtension(filePath),
    ["beVersion"] = root.Attribute("be_version")?.Value,
    ["framework"] = NodeText(root, "framework"),
};

// Languages
var languages = new List<string>();
var langsNode = root.Element("languages");
if (langsNode is not null)
{
    foreach (var l in langsNode.Elements("language"))
    {
        var code = NodeText(l, "code");
        if (code is not null)
            languages.Add(code);
    }
}
else
{
    // Fall back to scanning <translation> nodes
    foreach (var tr in root.Descendants("translation"))
    {
        var lang = NodeText(tr, "language");
        if (lang is not null && !languages.Contains(lang))
            languages.Add(lang);
    }
}
languages.Sort(StringComparer.Ordinal);
project["languages"] = languages;

// Packages
var packages = new List<Dictionary<string, object?>>();
foreach (var pkg in root.Descendants("package_node"))
{
    packages.Add(ParsePackage(pkg));
}
project["packages"] = packages;

// Translation packages
var translationPackages = new List<Dictionary<string, object?>>();
var tpNode = root.Element("translation_packages");
if (tpNode is not null)
{
    foreach (var tpackage in tpNode.Elements("translation_package"))
    {
        var tname = NodeText(tpackage, "name") ?? "";
        var turls = new List<Dictionary<string, string?>>();
        var urlsNode = tpackage.Element("translation_urls");
        if (urlsNode is not null)
        {
            foreach (var url in urlsNode.Elements("translation_url"))
            {
                turls.Add(new Dictionary<string, string?>
                {
                    ["path"] = NodeText(url, "path"),
                    ["language"] = NodeText(url, "language")
                });
            }
        }
        translationPackages.Add(new Dictionary<string, object?>
        {
            ["name"] = tname,
            ["translationUrls"] = turls
        });
    }
}
project["translationPackages"] = translationPackages;

// Embedded source texts & primary language
var est = NodeText(root, "embedded_source_texts");
project["embeddedSourceTexts"] = string.Equals(est ?? "false", "true", StringComparison.OrdinalIgnoreCase);
project["primaryLanguage"] = NodeText(root, "primary_language");

// Editor configuration
var editorCfg = new Dictionary<string, object>();
var edNode = root.Element("editor_configuration");
if (edNode is not null)
{
    foreach (var c in edNode.Elements())
    {
        if (c.Name.LocalName == "copy_template")
        {
            if (!editorCfg.TryGetValue("copy_templates", out var existing))
            {
                existing = new List<string>();
                editorCfg["copy_templates"] = existing;
            }
            ((List<string>)existing).Add(c.Value);
        }
        else
        {
            editorCfg[c.Name.LocalName] = c.Value;
        }
    }
}
project["editorConfiguration"] = editorCfg;

// Configuration
var cfg = new Dictionary<string, string?>();
var confNode = root.Element("configuration");
if (confNode is not null)
{
    foreach (var c in confNode.Elements())
    {
        cfg[c.Name.LocalName] = c.Value;
    }
}
project["configuration"] = cfg;

// Output
var options = new JsonSerializerOptions
{
    WriteIndented = true,
    DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping
};
Console.WriteLine(JsonSerializer.Serialize(project, options));
return 0;

// --- Helper methods ---

static string? NodeText(XElement parent, string childName)
{
    return parent.Element(childName)?.Value;
}

static Dictionary<string, object?> ParseTranslations(XElement? transElem)
{
    var result = new Dictionary<string, object?>();
    if (transElem is null) return result;

    foreach (var tr in transElem.Elements("translation"))
    {
        var lang = NodeText(tr, "language");
        if (lang is null) continue;

        var approvedText = NodeText(tr, "approved");
        var approved = string.Equals(approvedText, "true", StringComparison.OrdinalIgnoreCase);
        result[lang] = new Dictionary<string, object> { ["approved"] = approved };
    }
    return result;
}

static Dictionary<string, object?> ParseConcept(XElement concept)
{
    return new Dictionary<string, object?>
    {
        ["name"] = NodeText(concept, "name"),
        ["description"] = NodeText(concept, "description") ?? "",
        ["comment"] = NodeText(concept, "comment") ?? "",
        ["translations"] = ParseTranslations(concept.Element("translations"))
    };
}

static Dictionary<string, object?> ParseFolder(XElement node)
{
    var name = NodeText(node, "name");
    var folders = new List<Dictionary<string, object?>>();
    var concepts = new List<Dictionary<string, object?>>();

    var children = node.Element("children");
    if (children is not null)
    {
        foreach (var c in children.Elements())
        {
            if (c.Name.LocalName == "folder_node")
                folders.Add(ParseFolder(c));
            else if (c.Name.LocalName == "concept_node")
                concepts.Add(ParseConcept(c));
        }
    }

    return new Dictionary<string, object?>
    {
        ["name"] = name,
        ["folders"] = folders,
        ["concepts"] = concepts
    };
}

static Dictionary<string, object?> ParseFile(XElement fileNode)
{
    var fileName = NodeText(fileNode, "name");
    var folders = new List<Dictionary<string, object?>>();

    var children = fileNode.Element("children");
    if (children is not null)
    {
        foreach (var ch in children.Elements())
        {
            if (ch.Name.LocalName == "folder_node")
            {
                folders.Add(ParseFolder(ch));
            }
            else if (ch.Name.LocalName == "concept_node")
            {
                // Top-level concept in a file — wrap as a folderless concept
                folders.Add(new Dictionary<string, object?>
                {
                    ["name"] = "",
                    ["folders"] = new List<Dictionary<string, object?>>(),
                    ["concepts"] = new List<Dictionary<string, object?>> { ParseConcept(ch) }
                });
            }
        }
    }

    return new Dictionary<string, object?>
    {
        ["name"] = fileName,
        ["folders"] = folders
    };
}

static Dictionary<string, object?> ParsePackage(XElement pkg)
{
    var name = NodeText(pkg, "name");
    var files = new List<Dictionary<string, object?>>();

    var children = pkg.Element("children");
    if (children is not null)
    {
        foreach (var child in children.Elements())
        {
            if (child.Name.LocalName == "file_node")
                files.Add(ParseFile(child));
        }
    }

    return new Dictionary<string, object?>
    {
        ["name"] = name,
        ["files"] = files
    };
}
